lib/ts/recipe/emailverification/api/implementation.ts

```diff
@@ -5,62 +5,228 @@ import { GeneralErrorResponse } from "../../../types";
 import { EmailVerificationClaim } from "../emailVerificationClaim";
 import SessionError from "../../session/error";
 import { getEmailVerifyLink } from "../utils";
+import { getUser } from "../../..";
+import Session from "../../session";
+import AccountLinking from "../../accountlinking/recipe";
+import { AccountLinkingClaim } from "../../accountlinking/accountLinkingClaim";
+import { SessionContainerInterface } from "../../session/types";
 
 export default function getAPIInterface(): APIInterface {
     return {
-        verifyEmailPOST: async function ({
-            token,
-            options,
-            session,
-            userContext,
-        }): Promise<
-            { status: "OK"; user: User } | { status: "EMAIL_VERIFICATION_INVALID_TOKEN_ERROR" } | GeneralErrorResponse
+        verifyEmailPOST: async function (
+            this: APIInterface,
+            { token, options, session, userContext }
+        ): Promise<
+            | { status: "OK"; user: User; newSession?: SessionContainerInterface }
+            | { status: "EMAIL_VERIFICATION_INVALID_TOKEN_ERROR" }
+            | GeneralErrorResponse
         > {
-            const res = await options.recipeImplementation.verifyEmailUsingToken({ token, userContext });
+            const verifyTokenResponse = await options.recipeImplementation.verifyEmailUsingToken({
+                token,
+                userContext,
+            });
 
-            if (res.status === "OK" && session !== undefined) {
-                try {
-                    await session.fetchAndSetClaim(EmailVerificationClaim, userContext);
-                } catch (err) {
-                    // This should never happen, since we've just set the status above.
-                    if ((err as Error).message === "UNKNOWN_USER_ID") {
-                        throw new SessionError({
-                            type: SessionError.UNAUTHORISED,
-                            message: "Unknown User ID provided",
-                        });
+            if (verifyTokenResponse.status === "EMAIL_VERIFICATION_INVALID_TOKEN_ERROR") {
+                return verifyTokenResponse;
+            }
+
+            let primaryUserIdThatTheAccountWasLinkedTo = await AccountLinking.getInstanceOrThrowError().createPrimaryUserIdOrLinkAccounts(
```

Comment:
There is a comment in the lucid chart:

> There is a race condidtion where post account link callback can be called twice. One when the user clicks on the email verification link which has a session and one when the email verify API is called with the same session. The race condition happens if both happens at the same time. But, it is very rare as this would require user to refresh the page manually while clicking on the email verfication link

How is this correct? Cause the email verify GET API never does account linking right? So the post account linking callback will not get called ever..